src/TelemetryChannels/ServerTelemetryChannel/Shared/AdaptiveSamplingTelemetryProcessor.cs

```diff
@@ -0,0 +1,265 @@
+            SamplingPercentageEstimatorSettings settings,
+            AdaptiveSamplingPercentageEvaluatedCallback callback,
+            ITelemetryProcessor next)
+        {
+            this.estimatorSettings = settings;
+            this.evaluationCallback = callback;
+
+            // make estimatortelemetry processor  work after sampling was done
+            this.estimatorProcessor = new SamplingPercentageEstimatorTelemetryProcessor(settings, this.SamplingPercentageChanged, next);
+            this.samplingProcessor = new SamplingTelemetryProcessor(this.estimatorProcessor);
+
+            this.initialSamplingPercentage = 100.0;
+        }
+
+        /// <summary>
+        /// Gets or sets initial sampling percentage applied at the start
+        /// of the process to dynamically vary the percentage.
+        /// </summary>
+        public double InitialSamplingPercentage
+        {
+            get
+            {
+                return this.initialSamplingPercentage;
+            }
+
+            set
+            {
+                // note: 'initial' percentage will affect sampling even 
+                // if it was running for a while
+                this.initialSamplingPercentage = value;
+                this.samplingProcessor.SamplingPercentage = this.initialSamplingPercentage;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets maximum rate of telemetry items per second
+        /// dynamic sampling will try to adhere to.
+        /// </summary>
+        public double MaxTelemetryItemsPerSecond
+        {
+            get
+            {
+                return this.estimatorSettings.MaxTelemetryItemsPerSecond;
+            }
+
+            set
+            {
+                this.estimatorSettings.MaxTelemetryItemsPerSecond = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets minimum sampling percentage that can be set 
+        /// by the dynamic sampling percentage algorithm.
+        /// </summary>
+        public double MinSamplingPercentage
+        {
+            get
+            {
+                return this.estimatorSettings.MinSamplingPercentage;
+            }
+
+            set
+            {
+                this.estimatorSettings.MinSamplingPercentage = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets maximum sampling percentage that can be set 
+        /// by the dynamic sampling percentage algorithm.
+        /// </summary>
+        public double MaxSamplingPercentage
+        {
+            get
+            {
+                return this.estimatorSettings.MaxSamplingPercentage;
+            }
+
+            set
+            {
+                this.estimatorSettings.MaxSamplingPercentage = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets duration of the sampling percentage evaluation 
+        /// interval in seconds.
+        /// </summary>
+        public int EvaluationIntervalSeconds
```

Comment:
How difficult would it be to implement reading TimeSpan from config? I'd prefer to have it strongly typed.